### ISSUE TITLE
View with extent ignores padding for contraint

### ISSUE THREAD

**Issue by taam** [openlayers/openlayers#13284]
**Describe the bug**
A view with `extent` and `padding` seems to ignore the padding for the extent constraint, for `showFullExtent` true or false. However I do not know, if that's a bug or intended behavior. (Tested with v6.12.0.)

**Expected behavior**
I did expect the padding to be honored, so I can zoom out fully using `fit` with the extent (and padding, without disabling the extent constraint). This used to work with older versions of OpenLayers.

**Comment by ahocevar** [comment]
@taam Can you please share your view configuration and the `fit()` call you tried?

**Comment by taam** [comment]
Sure, I'll try, looks like that:

```js
var projection = new ol.proj.Projection({
	code: 'pixels',
	units: 'pixels',
	extent: extent,
});

var view = new ol.View({
	projection: projection,
	extent: extent,
	showFullExtent: true,
	center: ol.extent.getCenter(extent),
	resolutions: this.viewResolutions, // yes, there are big numbers as well
	zoom: 0,
	padding: [10, 20, 30, 40], // just testing
});

// [... creating map ...]

view.fit(extent, {
	padding: [10, 20, 30, 40], // just testing
	constrainResolution: false, // not required for v6?
});
```

It kinda works when enabling `constrainOnlyCenter`, but I actually want the constraint, just with the padding. I can probably cheat and extend the view extent, but that would be bad when zooming in or when changing the map size, so that does not seem like a good idea.

**Comment by ahocevar** [comment]
What happens if you remove the `padding` from the `fit()` options? That's redundant, because already configured on the `View` level.

**Comment by taam** [comment]
The `padding`  from the `fit()` options is actually behaving quite strange. I initially tested with symmetric padding, where it does not seem to make any difference indeed. But setting e.g. [500, 0, 0, 0] leads to a zoomed in state, where some of the top and the bottom(!) is out of the viewport. In any case, when zooming out using the mouse or the controls, it zooms out to the full extent, centered, but without padding.

(The actual code normally does `view.setCenter(...)` after the fit, but it seems not to matter regarding extent constraint.)

**Comment by taam** [comment]
For testing, I built a modified version of OpenLayers v6.12.0, where I replaced pretty much every call of `getViewportSize_` in `View.js` with `getViewportSizeMinusPadding_`. This seems to work for my specific use case (even though with rotation it's still not ideal, but that's probably another topic), but these changes might o.c. break something else.

**Comment by ahocevar** [comment]
I see now what you expect to happen, and I think it's a valid question to ask whether extent constraints should be considered for padding and fit. Originally, the view `padding` was meant to be a convenience configuration to avoid having to use `fit()` with an asymmetric `padding` when a UI element covers a substantial part of the map viewport. Similar to what Google Maps does when the side panel with search results et al. is visible. The idea was to set the padding when that UI element is expanded, and to unset the padding when it is collapsed.

So before committing to any changes, I'd like to ask the following:
1. Without a `padding` set on the view, does `fit()` behave as you'd expect?
2. Does your use case for a view `padding` match the above? If not, what's your use case?

**Comment by taam** [comment]
> 1. Without a `padding` set on the view, does `fit()` behave as you'd expect?

Yes. (At least I think so, I use it only for zooming out to the full extent.)

> 2. Does your use case for a view `padding` match the above? If not, what's your use case?

Not exactly, my padding is static, however, the result should be the same I guess?

(My use case is to show a border around an image, that represents the full extent. I set a small symmetric padding, which I think helps recognizing the border and therefore the edge of the image, more of a visual aid. Also, when enabling `showFullExtent`, to me it looks weird when the image is fully zoomed out and clamped in the viewport on two sides.)

However, the problem seems not to be the `fit()`, but the combination of `extent` constraint and `padding`. Even without `fit()`, it's not possible to zoom out e.g. via scroll wheel or buttons.

**Comment by taam** [comment]
I admit constraining the `extent` to a padded viewport might be not as easy with dynamic `padding`. One could consider it different use cases I guess...

**Comment by dennis-leee** [comment]
met this problem too. and my requirement is that
1. load some data from backend
2. render them to Map
3. limit the View with the init data's Extent, and the View has padding

and the problem is that, once i set the Extent in View, then the padding will be ignore.

then i try
1. use `View.fit` to let View fit to the Extent with a padding.
2. use `View.calculateExtent` to get a new Extent which include the padding
3. create a new View with the new Extent ( another question: why there no setExtent in View ?)

In most cases, the above approach can meet my requirement, but not the following case
1. user's browser window is small (like window mode) when init the map.
2. user make the browser window full screen
in this case, the extent that get from `View.calculateExtent` is not suitable for the enlarged window.
and user can't see all features in one screen of the enlarged window due to the "small" Extent limit

dut to the business case, i can't use `View.calculateExtent` to get a new Extent with padding to reset View after init map.
i think there is two way to solve this problem:

1. get a extent with padding when init map.
2. View compatible with the situation where Padding and Extent exist at the same time, dont ignore Padding.

for the first one, i found `buffer` method from `Extent`, but i can't find example for it.
and my case is that
1. `Extent` is in units of LngLat
2. `size` is in units of Pixel
should i do a transform when use it? if need, how?

for the second one, that's why i found this issue
